crates/ethereum/src/lib.rs

```diff
@@ -1,356 +1,214 @@
-            x if x == U256::from(11155111u32) => EthereumChain::Sepolia,
-            x => EthereumChain::Other(x),
-        })
-    }
-}
-
-fn encode_ethereum_call_data(signature: &[u8]) -> String {
-    let mut output: [u8; 32] = Default::default();
-    keccak_hash::keccak_256(signature, &mut output[..]);
-    format!("0x{}", hex::encode(&output[0..4]))
-}
-
-fn get_h256(value: &serde_json::Value) -> anyhow::Result<H256> {
-    use std::str::FromStr;
-    value
-        .as_str()
-        .map(lpad64)
-        .and_then(|val| H256::from_str(&val).ok())
-        .context("Failed to fetch H256")
-}
-
-fn get_u256(value: &serde_json::Value) -> anyhow::Result<U256> {
-    use std::str::FromStr;
-    value
-        .as_str()
-        .map(lpad64)
-        .and_then(|val| U256::from_str(&val).ok())
-        .context("Failed to fetch U256")
-}
-
-fn get_felt(value: H256) -> anyhow::Result<Felt> {
-    let felt = Felt::from_be_slice(value.as_bytes())?;
-    Ok(felt)
-}
-
-fn get_number(value: U256) -> anyhow::Result<BlockNumber> {
-    let value = value.as_u64();
-    BlockNumber::new(value).context("Failed to read u64 from U256")
-}
+    /// Listens for Ethereum events and notifies the caller using the provided
+    /// callback
+    async fn listen<F, Fut>(&self, address: &H160, callback: F) -> anyhow::Result<()>
+    where
+        F: Fn(EthereumEvent) -> Fut + Send + 'static,
+        Fut: Future<Output = ()> + Send + 'static,
+    {
+        // Create a WebSocket connection
+        let ws = WsConnect::new(self.url.clone());
+        let provider = ProviderBuilder::new().on_ws(ws).await?;
+
+        // Create the StarknetCoreContract instance
+        let address = Address::new((*address).into());
+        let core_contract = StarknetCoreContract::new(address, provider.clone());
+
+        // Listen for L1 to L2 message events
+        let mut logs = provider
+            .subscribe_logs(&core_contract.LogMessageToL2_filter().filter)
+            .await?
+            .into_stream();
+
+        // Listen for state update events
+        let mut state_updates = provider
+            .subscribe_logs(&core_contract.LogStateUpdate_filter().filter)
+            .await?
+            .into_stream();
+
+        loop {
+            select! {
+                Some(state_update) = state_updates.next() => {
+
+                    // Decode the state update
+                    let state_update: Log<StarknetCoreContract::LogStateUpdate> = state_update.log_decode()?;
+                    let state_update = EthereumStateUpdate {
+                        block_number: get_block_number(state_update.inner.blockNumber),
+                        block_hash: get_block_hash(state_update.inner.blockHash),
+                        state_root: get_state_root(state_update.inner.globalRoot),
+                    };
+
+                    // Emit the state update
+                    callback(EthereumEvent::StateUpdate(state_update)).await;
+
+                }
```

Comment:
I don't think we actually make sure that we only ever call the callback with finalized logs.

What we should be doing here is a bit more complicated:

- Store all events temporarily in-memory.
- Poll Ethereum for the latest finalized block number. If the finalized block number is updated we should filter the in-memory store for all pending events that have been finalized and emit those (and remove them from the in-memory store).
- We should handle the `removed` flag as well (`state_update.removed`): in case a removed log is received we should remove it from the in-memory store of pending events.